ipalib/util.py

```diff
@@ -187,6 +194,72 @@ def normalize_zone(zone):
         return zone
 
 
+class IPAHTTPSConnection(HTTPSConnection, object):
+    """
+    This class inherits from `object` because HTTPSConnection does not in
+    Python 2.7. This is to allow the use of super() in its and derived
+    classes' methods.
+    """
+
+    # pylint: disable=no-member
+    tls_cutoff_map = {
+        "tls1.0": ssl.OP_NO_TLSv1,
+        "tls1.1": ssl.OP_NO_TLSv1_1,
+        "tls1.2": ssl.OP_NO_TLSv1_2,
+    }
+    # pylint: enable=no-member
+
+    def __init__(self, host, port=HTTPSConnection.default_port, cafile=None,
+                 client_certfile=None, client_keyfile=None,
+                 tls_version_min="tls1.1",
+                 tls_version_max="tls1.2",
+                 **kwargs):
+        """
+        Set up a client HTTPS connection.
+
+        :param host:  The host to connect to
+        :param port:  The port to connect to, defaults to
+                   HTTPSConnection.default_port
+        :param cafile:  A PEM-format file containning the trusted
+                        CA certificates
+        :param client_certfile:
+                A PEM-format client certificate file that will be used to
+                identificate the user to the server. It should also contain
+                the client private key.
+        :returns An established HTTPS connection to host:port
+        """
+        # pylint: disable=no-member
+        tls_cutoff = [
+            ssl.OP_NO_TLSv1,
+            ssl.OP_NO_TLSv1_1,
+            ssl.OP_NO_TLSv1_2
+        ]
+        # remove the slice of negating protocol options according to options
+        min_idx = tls_cutoff.index(self.tls_cutoff_map[tls_version_min])
+        max_idx = tls_cutoff.index(self.tls_cutoff_map[tls_version_max])
+        del(tls_cutoff[min_idx:max_idx+1])
+
+        # official Python documentation states that the best option to get
+        # TLSv1 and later is to setup SSLContext with PROTOCOL_SSLv23
+        # and then negate the insecure SSLv2 and SSLv3
+        ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
```

Comment:
You want to set secure settings for cipher suites, too. The default settings are rather bad. In fact it would be better if you use ```ssl.create_default_context``` instead of ```ssl.SSLContext```.